### PR TITLE
- allow "exec" commmand to pass dictionary-params to webworker (with testcase)

### PR DESCRIPTION
**Pull request by kaizhu256**
- add test passing dictionary-params to webworker

i generally use sql.js via webworkers, b/c i need the browser-ui to be responsive while ingesting/manipulating largish (200mb) csv's and queries that could take 30+ seconds.

the biggest missing feature for me using webworkers is passing parameters.  i've been successfully using the patch in this pull-request in production for the past 6 months.

### CODE REVIEW

**Comment by kaizhu256** [comment]
added documentation for multi-line exec-statements with params in commit # 2: 7bdff4d.

also verified documented example-usage runs correctly in browser:

image

File: README.md [hunk 1]
```diff
@@ -204,15 +204,34 @@ Example:
...
+        "$id1": 1,
+        ":age1": 1,
+        "@name1": "Ling",
+        "$id2": 2,
+        ":age2": 18,
+        "@name2": "Paul",
+        "$id3": 3,
+        ":age3": 3,
+        "@name3": "Marcus"
+      }
```
Comment (lovasoa) [inline README.md]:
```suggestion
      sql: "SELECT age,name FROM test WHERE id=$id",
      params: { "$id": 1 }
```

File: src/api.js [hunk 2]
```diff
@@ -660,7 +661,8 @@ Module["onRuntimeInitialized"] = function onRuntimeInitialized() {
     * by a semicolon)
     *
     * ## Example use
-    * We have the following table, named *test* :
+    * We will create following table, named *test* and query it with a
```
Comment (lovasoa) [inline src/api.js]:
```suggestion
    * We will create the following table, named *test* and query it with a
```

File: src/api.js [hunk 3]
```diff
@@ -671,18 +673,44 @@ Module["onRuntimeInitialized"] = function onRuntimeInitialized() {
...
+    *         ":age1": 1,
+    *         "@name1": "Ling",
+    *         "$id2": 2,
+    *         ":age2": 18,
+    *         "@name2": "Paul",
+    *         "$id3": 3,
+    *         ":age3": 3,
+    *         "@name3": "Marcus"
+    *     }
+    * );
```
Comment (lovasoa) [inline src/api.js]:
```suggestion
    * var res = db.exec(
    *            "CREATE TABLE test (id INTEGER, age INTEGER, name TEXT);"
    *         + "INSERT INTO test VALUES ($id1, :age1, @name1); "
    *         + "SELECT age,name FROM test WHERE id=$id1",
    *         { "$id1": 1, ":age1": 1, "@name1": "Ling",  }
    * );
```

File: src/api.js (hunk 2 above)
Comment (kaizhu256) [reply src/api.js]: applied suggestion in commit 6e5024ed6c522e2719bb37e9418f708e29cf3786

File: examples/GUI/index.html [hunk 4]
```diff
@@ -49,7 +49,7 @@ <h1>Online SQL interpreter</h1>
 
     <button id="execute" class="button">Execute</button>
     <button id='savedb' class="button">Save the db</button>
-    <label class="button">Load an SQLite database file: <input type='file' id='dbfile'></label>
+    <label class="button">Load a SQLite database file: <input type='file' id='dbfile'></label>
```
Comment (lovasoa) [inline examples/GUI/index.html]: I think both are correct... The creator of SQLite says "an SQLite database"...

https://en.wikipedia.org/wiki/SQLite
https://english.stackexchange.com/questions/431329/what-is-the-correct-pronunciation-of-sqlite

File: src/api.js (hunk 3 above)
Comment (kaizhu256) [reply src/api.js]: the documented example does not produce the documented result.
if intent is to have a shorter, nicer looking example, then i compromised with following in commit 6e5024ed6c522e2719bb37e9418f708e29cf3786 if its ok:

```diff
     * multi-line statement using params:
     *
     * | id | age |  name  |
     * |:--:|:---:|:------:|
     * | 1  |  1  | Ling   |
     * | 2  |  18 | Paul   |
-    * | 3  |  3  | Markus |
     *
     * We query it like that:
     * ```javascript
     * var db = new SQL.Database();
     * var res = db.exec(
-    *     (
-    *         "DROP TABLE IF EXISTS test;\n"
-    *         + "CREATE TABLE test (id INTEGER, age INTEGER, name TEXT);\n"
-    *         + "INSERT INTO test VALUES ($id1, :age1, @name1);\n"
-    *         + "INSERT INTO test VALUES ($id2, :age2, @name2);\n"
-    *         + "INSERT INTO test VALUES ($id3, :age3, @name3);\n"
-    *         + "SELECT id FROM test;\n"
-    *         + "SELECT age,name FROM test;\n"
-    *     ),
+    *     "DROP TABLE IF EXISTS test;\n"
+    *     + "CREATE TABLE test (id INTEGER, age INTEGER, name TEXT);"
+    *     + "INSERT INTO test VALUES ($id1, :age1, @name1);"
+    *     + "INSERT INTO test VALUES ($id2, :age2, @name2);"
+    *     + "SELECT id FROM test;"
+    *     + "SELECT age,name FROM test WHERE id=$id1",
     *     {
-    *         "$id1": 1,
-    *         ":age1": 1,
-    *         "@name1": "Ling",
-    *         "$id2": 2,
-    *         ":age2": 18,
-    *         "@name2": "Paul",
-    *         "$id3": 3,
-    *         ":age3": 3,
-    *         "@name3": "Marcus"
+    *         "$id1": 1, ":age1": 1, "@name1": "Ling",
+    *         "$id2": 2, ":age2": 18, "@name2": "Paul"
     *     }
     * );
     * ```
@@ -700,8 +689,8 @@ Module["onRuntimeInitialized"] = function onRuntimeInitialized() {
     * `res` is now :
     * ```javascript
     *     [
-    *         {"columns":["id"],"values":[[1],[2],[3]]},
-    *         {"columns":["age","name"],"values":[[1,"Ling"],[18,"Paul"],[3,"Marcus"]]}
+    *         {"columns":["id"],"values":[[1],[2]]},
+    *         {"columns":["age","name"],"values":[[1,"Ling"]]}
     *     ]
     * ```
```

image

Comment (lovasoa) [reply src/api.js]: This seems reasonable to me 👍

File: src/api.js [hunk 5]
```diff
@@ -660,32 +661,48 @@ Module["onRuntimeInitialized"] = function onRuntimeInitialized() {
...
-    *         {columns: ['id'], values:[[1],[2],[3]]},
-    *         {columns: ['age','name'], values:[[1,'Ling'],[18,'Paul'],[3,'Markus']]}
+    *         {"columns":["id"],"values":[[1],[2]]},
+    *         {"columns":["age","name"],"values":[[1,"Ling"]]}
     *     ]
     * ```
     *
-    * @param {string} sql a string containing some SQL text to execute
+    @param {string} sql a string containing some SQL text to execute
+    @param {any[]} [params=[]] When the SQL statement contains placeholders,
```
Comment (lovasoa) [inline src/api.js]:
```suggestion
    @param {Statement.BindParams} [params=[]] When the SQL statement contains placeholders,
```

File: src/api.js [hunk 6]
```diff
@@ -660,32 +661,48 @@ Module["onRuntimeInitialized"] = function onRuntimeInitialized() {
...
+    you can pass them in here. They will be bound to the statement
+    before it is executed. If you use the params argument as an array,
+    you **cannot** provide a sql string that contains several queries
```
Comment (lovasoa) [inline src/api.js]:
```suggestion
    you **cannot** provide a sql string that contains several statements
```

File: examples/GUI/index.html (hunk 4 above)
Comment (kaizhu256) [reply examples/GUI/index.html]: ok reverted grammar-changes in commit c6aa8884baa921e3aedf31f30796db1ddece5c35

File: src/api.js (hunk 5 above)
Comment (kaizhu256) [reply src/api.js]: applied suggestion in commit f2d5ad2c1c5e0c3093c69811bf8ddc772413736c

File: src/api.js (hunk 6 above)
Comment (kaizhu256) [reply src/api.js]: applied suggestion in commit f2d5ad2c1c5e0c3093c69811bf8ddc772413736c

**Comment by kaizhu256** [comment]
log of commits-notes:

```
- allow "exec" commmand to pass dictionary-params to webworke
- add test passing dictionary-params to webworker
- add documentation
- resolve conversation - Statement.BindParams
- resolve conversation - replace "several queries" with "several statements"
- try to limit documentation to 80-column width
```

1 remaining unresolved conversation: example use-case @ line 178 in src/api.js

**Comment by lovasoa** [comment]
I don't see any unresolved conversation. Is this ready for merging?

**Comment by kaizhu256** [comment]
yes, its ready for merge on my side! : )

i have another worker patch in the works (to ingest large csv's while keeping ui-responsive) that depends on this.

**Comment by lovasoa** [comment]
Ok, it's merged. Thank you @kaizhu256. What is your other patch ? Maybe we can discuss it first ? I was thinking about deprecating our buggy and largely untested worker code in favor of more modern and feature-complete alternatives.

**Pull request merged**